### ISSUE TITLE
Error emitting signal: dbus: connection closed by user

### ISSUE THREAD

**Issue by Hjdskes** [godbus/dbus#109]
Hi,

I am trying to wrap a library to expose it over DBus. The library provides functionality to query and set Ikea Trådfri light bulbs, and also provides a channel over which notifications are sent when devices are changed through other means (such as a smartphone app or the remote control) so that any user of the library can update its state accordingly.

In my code, I spawn a goroutine which reads from this channel. For every received message I would then like to emit a notification, with the received message as value. The DBus daemon works in that I can query and set light bulbs through e.g. DFeet, but when I press any button on my remote in order to test the signal delivery I get `Error emitting signal: dbus: connection closed by user`. The provided channel works appropriately in a command line tool accompanying the library. I've been stuck on this for a while and cannot find a clue as to why this is happening, hence this issue.

The relevant code is as follows:

```
        ...

	err = d.conn.Export(DBusClient{*d.client}, objectRoot, nameRoot)
	if err != nil {
		return fmt.Errorf("Cannot export Trådfri interface: %s\n", err)
	}

	// Request name after exporting interfaces to ensure that the interfaces aren't called before they
	// are exported.
	reply, err := d.conn.RequestName(nameRoot, dbus.NameFlagDoNotQueue)
	if err != nil {
		return fmt.Errorf("Cannot associate name \"%s\": %s\n", nameRoot, err)
	}
	if reply != dbus.RequestNameReplyPrimaryOwner {
		return fmt.Errorf("Name \"%s\" already taken", nameRoot)
	}

        ...

        go d.handleDeviceUpdates()
```

Where `handleDeviceUpdates` is:

```
func (d *DBusDaemon) handleDeviceUpdates() {
	for msg := range d.client.DeviceEvents() {
		err := d.conn.Emit(objectRoot, nameRoot+".DeviceChanged", msg)
		if err != nil {
			fmt.Printf("Error emitting signal: %s\n", err)
		} else {
			fmt.Printf("Emitted signal %s.%s\n", nameRoot+".DeviceChanged", msg.ID)
		}
        }
}
```

The full code is [here](https://github.com/Hjdskes/tradfri). Any comments or pointers are appreciated!

**Comment by danields761** [comment]
Hi,

@Hjdskes, happened to me too. I found that dbus closes connection when it receives invalid dbus object path in any kind of cases: method call, method return and signals. 
Maybe it helps you :)

**Comment by Hjdskes** [comment]
Thanks for your reply @danields761! I just tried it again, and now (without changes) the behavior is a little different:

`d.conn.Emit` returns a `nil` error value exactly once; any successive call returns an error value indicating the bus has been closed again. However, `dbus-monitor` does not receive any signals. The only messages I see are the following:

```
signal time=1529263000.883889 sender=org.freedesktop.DBus -> destination=:1.118 serial=7 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameLost
   string "nl.hjdskes.Tradfri"
signal time=1529263000.883945 sender=org.freedesktop.DBus -> destination=(null destination) serial=209 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameOwnerChanged
   string "nl.hjdskes.Tradfri"
   string ":1.118"
   string ""
```

For good measure, I made the `objectRoot` constant a proper `ObjectPath`. Signals do not have to be "consumed" by anything, so what else can cause dbus to close the bus after emitting a signal?

**Comment by danields761** [comment]
Hi,

Probably the problem is that some objects are serialized with an errors, that's why dbus-daemon closes connection after unsuccessful decode attempt.
I dont know how to debug that because i'm unfamiliar with dbus binary protocol and with dbus-daemon sources. @jsouthworth, @guelfey, do you have any ideas?

Also this issue #85 is about the same problem, at least scenario looks same on 99%

**Comment by Hjdskes** [comment]
> Probably the problem is that some objects are serialized with an errors, that's why dbus-daemon closes connection after unsuccessful decode attempt.

This is the issue! When I only give the device's identifier (a uint32) the bus is not closed. I'll check my objects. I'll leave this issue open for now in case you want to do something with it (it would be nice if there would be a better error message indicating why the bus is closed, if that is possible).

**Comment by jsouthworth** [comment]
Can you provide an example of what was being encoded when the connection was closed? That should help isolate what isn't being encoded properly. Thanks.

**Comment by Hjdskes** [comment]
I was encoding the below struct using the DBus type `((ssssyy)a(siiiiyyiidudsu)ysxuyxi)`. I believe I tried to encode one of the integers wrongly, i.e. I was encoding a `uint32` as an `int` or some variation of that.

```
type DeviceDescription struct {
	Device struct {
		Manufacturer string
		ModelNumber string
		Serial string
		FirmwareVersion string
		AvailablePowerSource uint8
		BatteryLevel uint8
	}
	LightControl []LightControl
	Type uint8
	Name string
	CreatedAt int64
	ID uint32
	Reachable uint8
	LastSeen int64
	OtaUpdateState int
}
```

I think the problem here is not per sé that something is being encoded incorrectly, it's that the only error message returned is that the bus is closed but not *why*.

**Comment by danields761** [comment]
Hi,

@jsouthworth, here is most trivial 100% working [example](https://play.golang.org/p/fv--xIPu8dp). Invalid ObjectPath in parametres list causes next read message method call to return such error: `read unix @->@/tmp/dbus-VfdH1Tcp0p: use of closed network connection`.
Only such scenario occurs for me, but, as i can see in @Hjdskes example, not only invalid ObjectPath causes such errors.

**Comment by jsouthworth** [comment]
It looks like there are several problems that are presenting the same way (getting disconnected from DBus). I have a fix for one of them, #85. I'm investigating the others.

**Comment by utezduyar** [comment]
I have been having the same problem and my wild guess worked out! 

```
	conn, err := dbus.SystemBus()
	if err != nil {
		fmt.Printf("Failed to open system bus: %s", err)
	}
	defer conn.Close()
	if err := conn.Emit("/foo/bar", "foo.bar.Baz"); err != nil {
		fmt.Printf("Failed to emit signal: %s", err)
	}
```

Has been giving me exact error and it turned out that the problem was due to deferred conn.Close. My guess is something is async down in the Emit that before we get a change to send the signal connection is being closed.

**Comment by guelfey** [comment]
So in the end, this boiled down to 3 different problems, all of which resulting in the bus daemon closing the connection and are resolved now as far as I can tell:
1. Array of struct encoding bug -> fixed in #85 
2. Invalid ObjectPath -> application error, but reported properly in #270
3. `defer conn.Close()` -> application error
